File: tools/gerar_cronograma.py
# -*- coding: utf-8 -*-
"""Gera o Cronograma de Implantação do SIGER® (.docx) — documento obrigatório.

Distribui as agendas conforme as horas (cobradas + bonificadas), com os macro
tópicos por etapa/treinamento, datas previstas (em dias úteis) e modalidade.
Pode receber agendas explícitas no YAML; caso contrário, monta o plano
automaticamente a partir dos módulos contratados.

Uso:
    python gerar_cronograma.py [data/cronograma.yaml]
"""
import os
import re
import sys
import datetime
from docx import Document
from docx.shared import Pt, RGBColor
from docx.oxml.ns import qn
from docx.oxml import OxmlElement
import _common as C

NAVY = RGBColor(0x1F, 0x4E, 0x78)
_MESES = ["janeiro", "fevereiro", "março", "abril", "maio", "junho", "julho",
          "agosto", "setembro", "outubro", "novembro", "dezembro"]


def shade_header(row, fill="1F4E78"):
    for cell in row.cells:
        tcPr = cell._tc.get_or_add_tcPr()
        shd = OxmlElement("w:shd"); shd.set(qn("w:val"), "clear"); shd.set(qn("w:fill"), fill)
        tcPr.append(shd)
        for p in cell.paragraphs:
            for r in p.runs:
                r.font.color.rgb = RGBColor(0xFF, 0xFF, 0xFF); r.font.bold = True


def _num(s):
    m = re.search(r"\d+(?:[.,]\d+)?", str(s or ""))
    return float(m.group(0).replace(",", ".")) if m else 0.0


def _fmt(x):
    x = float(x or 0)
    return str(int(x)) if x == int(x) else ("%.1f" % x)


def _parse_date(s):
    s = str(s or "").strip()
    for fmt in ("%Y-%m-%d", "%d/%m/%Y", "%d/%m/%y"):
        try:
            return datetime.datetime.strptime(s, fmt).date()
        except ValueError:
            pass
    return datetime.date.today()


def _prox_util(d):
    while d.weekday() >= 5:           # 5=sáb, 6=dom
        d += datetime.timedelta(days=1)
    return d


def _add_uteis(d, n):
    while n > 0:
        d += datetime.timedelta(days=1)
        if d.weekday() < 5:
            n -= 1
    return d


def _distribuir(total, pesos):
    """Reparte `total` horas pelos pesos (método do maior resto), em inteiros.
    Sem total informado, usa peso*2 como horas."""
    if total <= 0:
        return [max(1, int(round(p * 2))) for p in pesos]
    soma = sum(pesos) or 1
    exatos = [total * p / soma for p in pesos]
    base = [int(x) for x in exatos]
    resto = int(round(total)) - sum(base)
    ordem = sorted(range(len(pesos)), key=lambda i: exatos[i] - base[i], reverse=True)
    i = 0
    while resto > 0 and ordem:
        base[ordem[i % len(ordem)]] += 1
        resto -= 1; i += 1
    return base


def _plano_automatico(modulos):
    """Plano padrão da implantação SIGER®: lista de (etapa, macro tópicos, peso)."""
    import catalogo as CAT
    achados, _ = CAT.resolve(modulos or [])
    plano = [
        ("Abertura + Parametrização inicial",
         "Criação de empresas e siglas; parâmetros gerais (1.1.P) e por empresa (1.2.A); "
         "compartilhamento de cadastros (1.2.M); tabelas genéricas.", 2.0),
    ]
    for m in achados:
        desc = m.get("descricao") or m.get("abrev") or "Módulo"
        plano.append(("Treinamento — %s" % desc,
                      "Tabelas e cadastros do módulo; importações via layout do SIGER®; "
                      "rotinas do processo; relatórios.", 2.0))
    if not achados:
        plano.append(("Treinamento das rotinas",
                      "Tabelas e cadastros; importações via layout; rotinas dos processos.", 4.0))
    plano += [
        ("Simulação de microprocessos",
         "Teste das rotinas treinadas por processo; ajustes finos.", 1.5),
        ("Simulação do macroprocesso",
         "Ensaio do processo completo (cenário real); validação ponta a ponta.", 1.5),
        ("Conversão — prévia",
         "Carga de teste; reconciliação origem × destino; validação de amostras.", 1.0),
        ("Conversão — oficial / ponto de corte",
         "Conversão oficial; conferência de saldos; definição do ponto de corte.", 1.0),
        ("Virada oficial (go-live)",
         "Início do uso em produção; acompanhamento full time; primeiros lançamentos.", 1.5),
        ("Acompanhamento / Hypercare",
         "Estabilização; micro ajustes; primeiros fechamentos.", 1.0),
        ("Encerramento",
         "Revisão de pendências; Termo de Encerramento; transição ao Suporte.", 0.5),
    ]
    return plano


def main(crono_path="data/cronograma.yaml"):
    d = C.load_yaml(os.path.basename(crono_path))
    cliente = d.get("cliente", "")
    num = d.get("numero_projeto", "")
    consultor = d.get("consultor", "")
    modalidade = d.get("modalidade_padrao", "A combinar")
    cadencia = int(_num(d.get("cadencia_dias", 5)) or 5)     # dias úteis entre agendas

    horas = d.get("horas") or {}
    cob = _num(horas.get("cobradas"))
    bon = _num(horas.get("bonificadas"))
    total = cob + bon or _num(horas.get("total") or d.get("horas_total"))

    # Agendas explícitas têm prioridade; senão, plano automático pelos módulos.
    agendas = d.get("agendas")
    if not agendas:
        plano = _plano_automatico(d.get("modulos_contratados") or d.get("modulos") or [])
        hs = _distribuir(total, [p for _, _, p in plano])
        if total <= 0:
            total = sum(hs)
        agendas = [{"etapa": e, "topicos": t, "horas": h}
                   for (e, t, _), h in zip(plano, hs)]

    # Datas previstas (dias úteis), espaçadas pela cadência.
    dt0 = _prox_util(_parse_date(d.get("data_inicio") or d.get("data")))
    for i, ag in enumerate(agendas):
        if not ag.get("data"):
            ag["data"] = (dt0 if i == 0 else _add_uteis(dt0, cadencia * i)).strftime("%d/%m/%Y")
        ag.setdefault("modalidade", modalidade)
        ag.setdefault("horas", "")
    soma_horas = sum(_num(a.get("horas")) for a in agendas)

    # --- Documento ---
    doc, based = C.style_base("cronograma")
    if not based:
        doc.styles["Normal"].font.name = "Calibri"
        doc.styles["Normal"].font.size = Pt(11)

    def P(txt="", bold=False):
        p = doc.add_paragraph(); p.add_run(txt).bold = bold
        return p

    C.docx_heading(doc, "Cronograma de Implantação do SIGER®", size=16, center=True)
    P("").add_run("Cliente: %s" % cliente).bold = True
    linha = []
    if num:
        linha.append("Projeto nº %s" % num)
    if consultor:
        linha.append("Consultor: %s" % consultor)
    linha.append("Horas: %s contratadas + %s bonificadas = %s"
                 % (_fmt(cob), _fmt(bon), _fmt(total)))
    P(" · ".join(linha))

    P("Este cronograma organiza a distribuição das agendas (horas contratadas + bonificadas), "
      "com os macro tópicos de cada visita/treinamento. As datas são previstas e podem ser "
      "ajustadas em comum acordo; o limite de 5 dias úteis após a liberação do levantamento é "
      "balizador. As agendas serão incluídas no SICLA (no nome do cliente) e compartilhadas com "
      "o usuário líder.")

    tab = doc.add_table(rows=1, cols=6); tab.style = "Table Grid"
    cabec = ["#", "Data prevista", "Etapa / Treinamento", "Macro tópicos", "Horas", "Modalidade"]
    for i, txt in enumerate(cabec):
        tab.rows[0].cells[i].paragraphs[0].add_run(txt)
    shade_header(tab.rows[0])
    for i, ag in enumerate(agendas, 1):
        c = tab.add_row().cells
        c[0].text = str(i)
        c[1].text = ag.get("data", "")
        c[2].text = ag.get("etapa", "")
        c[3].text = ag.get("topicos", "")
        h = _num(ag.get("horas"))
        c[4].text = (_fmt(h) + " h") if h else ""
        c[5].text = ag.get("modalidade", "")
    tot = tab.add_row().cells
    tot[3].paragraphs[0].add_run("Total").bold = True
    tot[4].paragraphs[0].add_run("%s h" % _fmt(soma_horas)).bold = True

    doc.add_paragraph()
    C.docx_heading(doc, "Observações", size=12)
    for it in (d.get("observacoes") or [
        "Datas previstas, sujeitas a ajuste conforme a disponibilidade das partes.",
        "Agendas registradas no SICLA no nome do cliente; treinamento/parametrização interna como tipo 84.",
        "Cronograma compartilhado via Google Drive (link enviado ao usuário líder).",
        "Cliente de menor porte: cronograma simplificado, deixando claro o que será e o que não será atendido.",
    ]):
        C.docx_bullet(doc, it)

    doc.add_paragraph()
    hoje = datetime.date.today()
    P(d.get("cidade_data") or ("Novo Hamburgo, %d de %s de %d."
                               % (hoje.day, _MESES[hoje.month - 1], hoje.year)))

    C.ensure_out()
    fname = "Cronograma_Implantacao_%s.docx" % C.slug(cliente)
    path = os.path.join(C.OUT, fname)
    doc.save(path)
    print("OK: %s -> %s" % (fname, path))


if __name__ == "__main__":
    args = sys.argv[1:]
    main(*args) if args else main()

File: backend/src/geradores/gerar-cronograma.ts
import { Document, Packer, Paragraph, TextRun } from 'docx';
import { existsSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { OUT_DIR, carregarYaml, slug } from './comum';
import { tabelaComCabecalho, tituloRech } from './comum-docx';
import { resolverModulos } from './catalogo';

/** Porte de `tools/gerar_cronograma.py` para Node/TS (§4.2/§4.7 dos Padrões da Rech).
 *
 * Gera o Cronograma de Implantação do SIGER® (.docx) — documento OBRIGATÓRIO. Distribui as
 * agendas conforme as horas (cobradas + bonificadas), com os macro tópicos por etapa, datas
 * previstas em dias úteis e modalidade. Porte de EQUIVALÊNCIA — prova em
 * `gerar-cronograma.spec.ts`.
 *
 * O original chama `_common.style_base("cronograma")`, que carregaria
 * `tools/templates/base_cronograma.docx` se ele existisse — e NÃO existe. Hoje o Python cai
 * no ramo de documento em branco, e é esse comportamento real que este porte reproduz. Se
 * algum dia o template for adicionado, o Python passa a usá-lo e este porte diverge; o teste
 * `gerar-cronograma.spec.ts` vigia exatamente isso e falha para avisar. */

const MESES = [
  'janeiro',
  'fevereiro',
  'março',
  'abril',
  'maio',
  'junho',
  'julho',
  'agosto',
  'setembro',
  'outubro',
  'novembro',
  'dezembro',
];

/** Caminho do template que o original procuraria — usado só pela guarda do teste. */
export const CAMINHO_TEMPLATE_BASE = join(
  process.cwd(),
  '..',
  'tools',
  'templates',
  'base_cronograma.docx',
);

export function templateBaseExiste(): boolean {
  return existsSync(CAMINHO_TEMPLATE_BASE);
}

interface Agenda {
  etapa?: string;
  topicos?: string;
  horas?: string | number;
  data?: string;
  modalidade?: string;
}
interface DadosCronograma {
  cliente?: string;
  numero_projeto?: string | number;
  consultor?: string;
  modalidade_padrao?: string;
  cadencia_dias?: string | number;
  horas?: {
    cobradas?: string | number;
    bonificadas?: string | number;
    total?: string | number;
  };
  horas_total?: string | number;
  agendas?: Agenda[];
  modulos_contratados?: (string | number)[];
  modulos?: (string | number)[];
  data_inicio?: string;
  data?: string;
  observacoes?: string[];
  cidade_data?: string;
}

/** Valor vindo do YAML que estes utilitários aceitam. Objeto aqui é bug, não uso — mesmo
 * critério já aplicado em `slug()` e `parseData()` de `comum.ts`. */
type ValorYaml = string | number | null | undefined;

/** Primeiro número do texto, com vírgula decimal aceita — `_num()`. */
export function num(s: ValorYaml): number {
  const m = /\d+(?:[.,]\d+)?/.exec(String(s ?? ''));
  return m ? parseFloat(m[0].replace(',', '.')) : 0;
}

/** Inteiro quando exato, senão uma casa decimal — `_fmt()`. */
export function fmt(x: number): string {
  const v = x || 0;
  return v === Math.trunc(v) ? String(Math.trunc(v)) : v.toFixed(1);
}

/** Aceita aaaa-mm-dd, dd/mm/aaaa e dd/mm/aa; cai para HOJE se não reconhecer — `_parse_date()`. */
export function parseDataCronograma(s: ValorYaml): Date {
  const texto = String(s ?? '').trim();
  let m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(texto);
  if (m) return new Date(+m[1], +m[2] - 1, +m[3]);
  m = /^(\d{2})\/(\d{2})\/(\d{4})$/.exec(texto);
  if (m) return new Date(+m[3], +m[2] - 1, +m[1]);
  m = /^(\d{2})\/(\d{2})\/(\d{2})$/.exec(texto);
  // %y do Python: 00-68 -> 2000-2068, 69-99 -> 1969-1999.
  if (m) {
    const aa = +m[3];
    return new Date(aa <= 68 ? 2000 + aa : 1900 + aa, +m[2] - 1, +m[1]);
  }
  return new Date();
}

/** Adianta para o próximo dia útil — `_prox_util()`. */
export function proximoUtil(d: Date): Date {
  const r = new Date(d);
  while (r.getDay() === 0 || r.getDay() === 6) r.setDate(r.getDate() + 1);
  return r;
}

/** Soma `n` dias ÚTEIS — `_add_uteis()`. */
export function somarUteis(d: Date, n: number): Date {
  const r = new Date(d);
  let faltam = n;
  while (faltam > 0) {
    r.setDate(r.getDate() + 1);
    if (r.getDay() !== 0 && r.getDay() !== 6) faltam -= 1;
  }
  return r;
}

function ddmmaaaa(d: Date): string {
  const p = (n: number) => String(n).padStart(2, '0');
  return `${p(d.getDate())}/${p(d.getMonth() + 1)}/${d.getFullYear()}`;
}

/** Reparte `total` horas pelos pesos (método do maior resto) — `_distribuir()`.
 * Sem total informado, usa peso*2 como horas. */
export function distribuir(total: number, pesos: number[]): number[] {
  if (total <= 0) return pesos.map((p) => Math.max(1, Math.round(p * 2)));
  const soma = pesos.reduce((a, b) => a + b, 0) || 1;
  const exatos = pesos.map((p) => (total * p) / soma);
  const base = exatos.map((x) => Math.trunc(x));
  let resto = Math.round(total) - base.reduce((a, b) => a + b, 0);
  const ordem = pesos
    .map((_, i) => i)
    .sort((a, b) => exatos[b] - base[b] - (exatos[a] - base[a]));
  let i = 0;
  while (resto > 0 && ordem.length > 0) {
    base[ordem[i % ordem.length]] += 1;
    resto -= 1;
    i += 1;
  }
  return base;
}

/** Plano padrão da implantação SIGER®: (etapa, macro tópicos, peso) — `_plano_automatico()`. */
export function planoAutomatico(
  modulos: (string | number)[] | undefined,
): [string, string, number][] {
  const { achados } = resolverModulos(modulos ?? []);
  const plano: [string, string, number][] = [
    [
      'Abertura + Parametrização inicial',
      'Criação de empresas e siglas; parâmetros gerais (1.1.P) e por empresa (1.2.A); ' +
        'compartilhamento de cadastros (1.2.M); tabelas genéricas.',
      2.0,
    ],
  ];
  for (const m of achados) {
    const desc = m.descricao || m.abrev || 'Módulo';
    plano.push([
      `Treinamento — ${desc}`,
      'Tabelas e cadastros do módulo; importações via layout do SIGER®; ' +
        'rotinas do processo; relatórios.',
      2.0,
    ]);
  }
  if (achados.length === 0) {
    plano.push([
      'Treinamento das rotinas',
      'Tabelas e cadastros; importações via layout; rotinas dos processos.',
      4.0,
    ]);
  }
  plano.push(
    [
      'Simulação de microprocessos',
      'Teste das rotinas treinadas por processo; ajustes finos.',
      1.5,
    ],
    [
      'Simulação do macroprocesso',
      'Ensaio do processo completo (cenário real); validação ponta a ponta.',
      1.5,
    ],
    [
      'Conversão — prévia',
      'Carga de teste; reconciliação origem × destino; validação de amostras.',
      1.0,
    ],
    [
      'Conversão — oficial / ponto de corte',
      'Conversão oficial; conferência de saldos; definição do ponto de corte.',
      1.0,
    ],
    [
      'Virada oficial (go-live)',
      'Início do uso em produção; acompanhamento full time; primeiros lançamentos.',
      1.5,
    ],
    [
      'Acompanhamento / Hypercare',
      'Estabilização; micro ajustes; primeiros fechamentos.',
      1.0,
    ],
    [
      'Encerramento',
      'Revisão de pendências; Termo de Encerramento; transição ao Suporte.',
      0.5,
    ],
  );
  return plano;
}

/** Monta o documento. Separado de `gerar()` para o teste inspecionar sem gravar. */
export function montarDocumento(
  d: DadosCronograma = carregarYaml<DadosCronograma>('cronograma.yaml'),
): Document {
  const cliente = d.cliente ?? '';
  const numeroProjeto = d.numero_projeto ?? '';
  const consultor = d.consultor ?? '';
  const modalidade = d.modalidade_padrao ?? 'A combinar';
  const cadencia = Math.trunc(num(d.cadencia_dias ?? 5)) || 5;

  const horas = d.horas ?? {};
  const cobradas = num(horas.cobradas);
  const bonificadas = num(horas.bonificadas);
  // O original é `cob + bon or _num(...)`: a soma só é descartada quando dá zero.
  let total = cobradas + bonificadas || num(horas.total ?? d.horas_total);

  // Agendas explícitas têm prioridade; senão, plano automático pelos módulos.
  let agendas: Agenda[] = d.agendas ?? [];
  if (agendas.length === 0) {
    const plano = planoAutomatico(d.modulos_contratados ?? d.modulos ?? []);
    const hs = distribuir(
      total,
      plano.map(([, , peso]) => peso),
    );
    if (total <= 0) total = hs.reduce((a, b) => a + b, 0);
    agendas = plano.map(([etapa, topicos], i) => ({
      etapa,
      topicos,
      horas: hs[i],
    }));
  }

  // Datas previstas (dias úteis), espaçadas pela cadência.
  const dt0 = proximoUtil(parseDataCronograma(d.data_inicio ?? d.data));
  agendas.forEach((ag, i) => {
    if (!ag.data) {
      ag.data = ddmmaaaa(i === 0 ? dt0 : somarUteis(dt0, cadencia * i));
    }
    ag.modalidade = ag.modalidade ?? modalidade;
    ag.horas = ag.horas ?? '';
  });
  const somaHoras = agendas.reduce((acc, a) => acc + num(a.horas), 0);

  const linha: string[] = [];
  if (numeroProjeto) linha.push(`Projeto nº ${numeroProjeto}`);
  if (consultor) linha.push(`Consultor: ${consultor}`);
  linha.push(
    `Horas: ${fmt(cobradas)} contratadas + ${fmt(bonificadas)} bonificadas = ${fmt(total)}`,
  );

  const observacoes = d.observacoes ?? [
    'Datas previstas, sujeitas a ajuste conforme a disponibilidade das partes.',
    'Agendas registradas no SICLA no nome do cliente; treinamento/parametrização interna como tipo 84.',
    'Cronograma compartilhado via Google Drive (link enviado ao usuário líder).',
    'Cliente de menor porte: cronograma simplificado, deixando claro o que será e o que não será atendido.',
  ];

  const hoje = new Date();
  const cidadeData =
    d.cidade_data ??
    `Novo Hamburgo, ${hoje.getDate()} de ${MESES[hoje.getMonth()]} de ${hoje.getFullYear()}.`;

  // Linha de total: só as colunas "Macro tópicos" e "Horas" são preenchidas, como no original.
  const linhasAgenda = agendas.map((ag, i) => {
    const h = num(ag.horas);
    return [
      String(i + 1),
      ag.data ?? '',
      ag.etapa ?? '',
      ag.topicos ?? '',
      h ? `${fmt(h)} h` : '',
      ag.modalidade ?? '',
    ];
  });
  linhasAgenda.push(['', '', '', 'Total', `${fmt(somaHoras)} h`, '']);

  return new Document({
    sections: [
      {
        children: [
          tituloRech('Cronograma de Implantação do SIGER®', 16, true),
          new Paragraph({
            children: [
              new TextRun({ text: `Cliente: ${cliente}`, bold: true }),
            ],
          }),
          new Paragraph(linha.join(' · ')),
          new Paragraph(
            'Este cronograma organiza a distribuição das agendas (horas contratadas + bonificadas), ' +
              'com os macro tópicos de cada visita/treinamento. As datas são previstas e podem ser ' +
              'ajustadas em comum acordo; o limite de 5 dias úteis após a liberação do levantamento é ' +
              'balizador. As agendas serão incluídas no SICLA (no nome do cliente) e compartilhadas com ' +
              'o usuário líder.',
          ),
          tabelaComCabecalho(
            [
              '#',
              'Data prevista',
              'Etapa / Treinamento',
              'Macro tópicos',
              'Horas',
              'Modalidade',
            ],
            linhasAgenda,
          ),
          new Paragraph({}),
          tituloRech('Observações', 12),
          ...observacoes.map(
            (it) => new Paragraph({ text: it, bullet: { level: 0 } }),
          ),
          new Paragraph({}),
          new Paragraph(cidadeData),
        ],
      },
    ],
  });
}

/** Gera o arquivo em `exemplos/`, com o mesmo nome do original. */
export async function gerar(): Promise<string> {
  const d = carregarYaml<DadosCronograma>('cronograma.yaml');
  const doc = montarDocumento(d);
  mkdirSync(OUT_DIR, { recursive: true });
  const caminho = join(
    OUT_DIR,
    `Cronograma_Implantacao_${slug(d.cliente)}.docx`,
  );
  writeFileSync(caminho, await Packer.toBuffer(doc));
  return caminho;
}

if (require.main === module) {
  gerar()
    .then((caminho) => console.log(`OK: ${caminho}`))
    .catch((e) => {
      console.error(e);
      process.exitCode = 1;
    });
}
